src/helics/common/SmallBuffer.hpp

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright (c) 2017-2020,
+Battelle Memorial Institute; Lawrence Livermore National Security, LLC; Alliance for Sustainable
+Energy, LLC.  See the top-level NOTICE for additional details. All rights reserved.
+SPDX-License-Identifier: BSD-3-Clause
+*/
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <string>
+#include <string_view>
+#include <utility>
+
+namespace helics {
+class SmallBuffer {
+  public:
+    SmallBuffer() noexcept: heap(buffer.data()) {}
+
+    SmallBuffer(const SmallBuffer& sb);
+    SmallBuffer(SmallBuffer&& sb) noexcept;
+    SmallBuffer(std::string_view val);
+    SmallBuffer(const void* data, size_t size);
+    ~SmallBuffer();
+    SmallBuffer& operator=(const SmallBuffer& sb);
+    SmallBuffer& operator=(SmallBuffer&& sb) noexcept;
+    SmallBuffer& operator=(std::string_view val);
+    /** return a pointer to the data location*/
+    std::byte* data() const { return heap; }
+    /** get the start of the data*/
+    std::byte* begin() { return heap; }
+    /** end iterator*/
+    std::byte* end() { return heap + bufferSize; }
+    /** get a const iterator*/
+    const std::byte* begin() const { return heap; }
+    /** get a const end iterator*/
+    const std::byte* end() const { return heap + bufferSize; }
+    /** get an element*/
+    std::byte operator[](size_t index) const { return heap[index]; }
```

Comment:
As mentioned elsewhere in this review, this operator does no validity checks, allowing unbounded access to elements outside the array. While it adds overhead, adding a safe array accessor would be preferable for safety.

Reply:
I think I would defer to std::vector here.  I added an `at` function which does the bounds checking and will throw a std::out_of_range exception.  Where as [] does no checking.